replica/table.cc

```diff
@@ -3495,20 +3524,24 @@ future<> compaction_group::cleanup() {
     co_await _t._cache.invalidate(std::move(updater), p_range);
 }
 
-future<> table::cleanup_tablet(database& db, db::system_keyspace& sys_ks, locator::tablet_id tid) {
+future<> table::cleanup_tablet(database& db, db::system_keyspace& sys_ks, locator::tablet_id tid, bool deallocate_sg) {
     auto holder = async_gate().hold();
 
-    auto& sg = storage_groups()[tid.value()];
-
-    for (auto& cg_ptr : sg->compaction_groups()) {
-        if (!cg_ptr) {
-            throw std::runtime_error(format("Cannot cleanup tablet {} of table {}.{} because it is not allocated in this shard",
-                                            tid, _schema->ks_name(), _schema->cf_name()));
-        }
+    auto& sg = _sg_manager->storage_groups()[tid.value()];
+    if (!sg) {
+        throw std::runtime_error(format("Cannot cleanup tablet {} of table {}.{} because it is not allocated in this shard",
+                                    tid, _schema->ks_name(), _schema->cf_name()));
+    }
+    utils::small_vector<compaction_group*, 3> compaction_groups = sg->compaction_groups();
+    dht::token_range range = sg->token_range();
 
+    // Keep the storage group alive.
+    std::unique_ptr<storage_group> sg_ptr = deallocate_sg ? std::exchange(_sg_manager->storage_groups()[tid.value()], nullptr) : nullptr;
```

Comment:
When we abort migration in stage allow_write_both_read_old we jump directly to revert_migration bypassing cleanup on the pending replica. If we want to rely on going via cleanup before deallocating replica, then we should change the rollback to always go through cleanup. I think it's fine, and safer. \cc @xemul  

See 6a77f3651959d342bd152eafaa82df1324adba81.